app/scripts/contentscript.js

```diff
@@ -64,18 +54,23 @@ let extensionMux,
 /**
  * Injects a script tag into the current document
  *
- * @param {string} content - Code to be executed in the current document
+ * @param {string} src - Path to code to be executed in the current document.
+ * @example injectScript(chrome.extension.getURL('inpage.js'))
  */
-function injectScript(content) {
+function injectScript(src) {
   try {
-    const container = document.head || document.documentElement;
-    const scriptTag = document.createElement('script');
-    scriptTag.setAttribute('async', 'false');
-    scriptTag.textContent = content;
-    container.insertBefore(scriptTag, container.children[0]);
-    container.removeChild(scriptTag);
+    const script = document.createElement("script");
+    // A script that has been injected into the DOM is executed asynchronously
+    // by default, but we need inpage.js to block so `window.ethereum` is
+    // available before the page's own scripts load.
+    script.async = false
```

Comment:
This isn't going to cut it.

`async=false` is only relevant in relations with other externally loaded scripts (aka `script:src`), but not with inline scripts (aka `script:textContent`). This means setting `async=false` will make sure the script runs before all other former externally loaded scripts, but former inline scripts that are part of the initial HTML will execute before it.

This will introduce a breaking change:
* `develop` before #22524 - our in-page script executes before any other script introduced in dApps, because it's being injected first as an inline script
* `develop` after #22524 - our in-page script executes before any other script introduced in dApps that is loaded externally - but not before inline scripts, because it's being injected first as an external `async=false` script.

This will effectively introduce a breaking change, where dApps attempting to access `window.ethereum` via inline scripts will move from succeeding to do so to failing.

see [example demo](https://weizmangal.com/public/demo-temp.html?silent&msg=111)

Reply:
Options:

1. `document.write()`

Browsing through https://github.com/whatwg/html/issues/7976, and especially [this X thread](https://twitter.com/zcorpan/status/1531675091657965572), one can learn that `document.write` is basically the only API allowing to introduce dynamic loading scripts that can block initial render.

So the following:

```javascript
function injectScript(src) {
  document.write(`
    <script id="xxxyyy" src="${src}"></script>
    <script>xxxyyy.remove(); document.currentScript.remove();</script>
  `);
}
```

should do the trick of both loading the script externally but truly synchronously by blocking the initial HTML rendering.

2. sync `XHR`

Another option is to fetch the contents of the in-page script synchronously and then load it as an inline script, thus making sure it runs before all other scripts to come.

```javascript
function injectScript(src) {
  const xhr = new XMLHttpRequest();
  xhr.open('GET', src, true);
  xhr.send();
  const content = xhr.responseText;
  const script = document.createElement('script');
  script.textContent = content;
  document.documentElement.prepend(script);
  script.remove();
}
```

3. hardcoded

We could also consider sticking to the already used method, which is probably somewhat better than these two above, but obviously you have your reasons for moving away from it, otherwise this PR wouldn't have existed.

Reply:
As for my recommendation,

Option[3] is probably the best. I don't understand the complications that drove you to introduce this change, but I wonder if they're worth it, because both the other two options will introduce some perf overhead.

As for option[1], I think `document.write` isn't as bad as is publicly considered, especially given our scenario.
https://csswizardry.com/2023/01/why-not-document-write/ goes in depth regarding `document.write` and my conclusion is that for our specific case `document.write` is a legit choice that shouldn't harm the dApps we inject ourselves into.

Option[2] is quite similar I guess, with also bad reputation (around sync XHR calls), but would get the job done similarly.

Problem with both is that they introduce more parser-blocking operations to the dApps parsing stage, as opposed to option[3]...